dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java

```diff
@@ -23,72 +23,136 @@
+    public void bindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        this.configFromServerXML = openAPIAppConfigProvider;
+        openAPIAppConfigProvider.registerAppConfigListener(this);
+    }
+
+    public void unbindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        openAPIAppConfigProvider.unregisterAppConfigListener(this);
+        this.configFromServerXML = null;
+    }
+
+    //This is here to stop injected trace calling lazyInit via toString when the class is constructed
+    @Trivial
+    public ModuleSelectionConfigImpl() {}
+
     /**
-     * Builds a {@code ModuleSelectionConfig} based on the given {@code Config}
+     * Builds a {@code ModuleSelectionConfig}
      * <p>
-     * This will read and parse the {@value Constants#MERGE_INCLUDE_CONFIG} and {@value Constants#MERGE_EXCLUDE_CONFIG} config properties.
+     * This will read and parse the {@value Constants#MERGE_INCLUDE_CONFIG} and {@value Constants#MERGE_EXCLUDE_CONFIG} config properties from mpConfig.
+     * <p>
+     * TODO doc what it reads from server.xml
      * <p>
      * If the config is invalid, this method will output warning messages but still return a usable result object.
      *
-     * @param config the config to read
      * @return the module selection config
      */
-    public static ModuleSelectionConfig fromConfig(Config config) {
-        ModuleSelectionConfig result = new ModuleSelectionConfig();
+    private synchronized void lazyInit() {
+        if (hasBeenInit) {
+            return;
+        }
 
-        String inclusion = config.getOptionalValue(Constants.MERGE_INCLUDE_CONFIG, String.class).orElse("first").trim();
+        Config configFromMPConfig = ConfigProvider.getConfig(ApplicationRegistryImpl.class.getClassLoader());
+
+        String inclusion;
+        if (!ProductInfo.getBetaEdition()) {
+            serverxmlmode = false;
+        } else {
+            serverxmlmode = configFromServerXML.getIncludedModules().isPresent() || configFromServerXML.getExcludedModules().isPresent();
+        }
+
+        if (serverxmlmode) {
+            //In another PR we will change first to a version specific service
+            inclusion = configFromServerXML.getIncludedModules().orElse("first").trim();
+        } else {
+            inclusion = configFromMPConfig.getOptionalValue(Constants.MERGE_INCLUDE_CONFIG, String.class).orElse("first");
+        }
 
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-            Tr.debug(null, tc, "Names in config: " + config.getPropertyNames());
+            if (serverxmlmode) {
+                Tr.debug(null, tc, "Acquired includes statement from server.xml");
+            } else {
+                Tr.debug(null, tc, "Names in config: " + configFromMPConfig.getPropertyNames());
+            }
             Tr.debug(null, tc, "Inclusion read from config: " + inclusion);
         }
```

Comment:
When these debug calls moved from a static method to a non-static method, you need to add the instance.

`Tr.debug(null, tc, ...)` becomes `Tr.debug(this, tc, ...)`